File: Anti-Fraud System/task/src/antifraud/util/CardNumberConstraint.java
package antifraud.util;


import jakarta.validation.Constraint;
import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import jakarta.validation.Payload;

import java.lang.annotation.Documented;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.*;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Target({FIELD, PARAMETER})
@Retention(RUNTIME)
@Constraint(validatedBy = CardNumberConstraintValidator.class)
@Documented
@Repeatable(CardNumberConstraint.List.class)
public @interface CardNumberConstraint {

    String message() default "{invalid card number}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

    @Target(FIELD)
    @Retention(RUNTIME)
    @Documented
    @interface List {
        CardNumberConstraint[] value();
    }
}


class CardNumberConstraintValidator implements ConstraintValidator<CardNumberConstraint, String> {
    @Override
    public void initialize(CardNumberConstraint constraintAnnotation) {
        ConstraintValidator.super.initialize(constraintAnnotation);
    }

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return CardNumberValidator.INSTANCE.isValid(value);
    }
}

File: Anti-Fraud System/task/src/antifraud/validation/util/CardNumberConstraint.kt
package antifraud.validation.util

import antifraud.validation.util.LuhnAlgorithm.checksum
import jakarta.validation.Constraint
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import jakarta.validation.Payload
import kotlin.reflect.KClass

private const val CARD_NUMBER_LENGTH = 16

@Target(AnnotationTarget.FIELD)
@MustBeDocumented
@Constraint(validatedBy = [CardNumberValidator::class])
annotation class CardNumberConstraint(
    val message: String = "{Invalid card number}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)

class CardNumberValidator : ConstraintValidator<CardNumberConstraint, String> {
    override fun isValid(value: String, context: ConstraintValidatorContext): Boolean {
        return isValidCardNumber(value)
    }
}

fun isValidCardNumber(cardNumber: String): Boolean {
    return if (cardNumber.length == CARD_NUMBER_LENGTH) {
        try {
            cardNumber.toLong()
            checksum(cardNumber)
        } catch (e: NumberFormatException) {
            false
        }
    } else false
}